packages/shell-api/src/bulk.ts

```diff
@@ -0,0 +1,217 @@
+import { hasAsyncChild, returnsPromise, ShellApiClass, shellApiClassDefault } from './decorators';
+import Mongo from './mongo';
+import { MongoshInternalError, MongoshInvalidInputError, MongoshUnimplementedError } from '@mongosh/errors';
+import {
+  Document,
+  WriteConcern
+} from '@mongosh/service-provider-core';
+import { assertArgsDefined } from './helpers';
+import { BulkWriteResult } from './result';
+
+@shellApiClassDefault
+export class BulkFindOp {
+  _innerFind: any;
+  _parentBulk: Bulk;
+  _hint: any;
+  _arrayFilters: any;
+  constructor(innerFind: any, parentBulk: Bulk) {
+    this._innerFind = innerFind;
+    this._parentBulk = parentBulk;
+  }
+
+  _asPrintable(): string {
+    return 'BulkFindOp';
+  }
+
+  // Blocked by NODE-2757
+  collation(): void {
+    throw new MongoshUnimplementedError(
+      'collation method on fluent Bulk API is not currently supported. ' +
+      'As an alternative, consider using the \'db.collection.bulkWrite(...)\' helper ' +
+      'which accepts \'collation\' as a field in the operations.'
+    );
+  }
+
+  // Blocked by NODE-2751
+  arrayFilters(): BulkFindOp {
+    throw new MongoshUnimplementedError(
+      'arrayFilters method on fluent Bulk API is not currently supported.'
+    );
+  }
+
+  hint(hintDoc: Document): BulkFindOp {
+    assertArgsDefined(hintDoc);
+    this._hint = hintDoc;
+    return this;
+  }
+
+  remove(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.remove();
+    return this._parentBulk;
+  }
+
+  removeOne(): Bulk {
+    this._parentBulk._batchCounts.nRemoveOps++;
+    this._innerFind.removeOne();
+    return this._parentBulk;
+  }
+
+  replaceOne(replacement: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(replacement);
+    const op = { ...replacement };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    this._innerFind.replaceOne(op);
+    return this._parentBulk;
+  }
+
+  updateOne(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.updateOne(op);
+    return this._parentBulk;
+  }
+
+  update(update: Document): Bulk {
+    this._parentBulk._batchCounts.nUpdateOps++;
+    assertArgsDefined(update);
+    const op = { ...update };
+    if (this._hint) {
+      op.hint = this._hint;
+    }
+    if (this._arrayFilters) {
+      op.arrayFilters = this._arrayFilters;
+    }
+    this._innerFind.update(op);
+    return this._parentBulk;
+  }
+
+  upsert(): BulkFindOp {
+    assertArgsDefined();
+    this._innerFind.upsert();
+    return this;
+  }
+}
+
+
+@shellApiClassDefault
+@hasAsyncChild
+export default class Bulk extends ShellApiClass {
+  _mongo: Mongo;
+  _collection: any; // to avoid circular ref
+  _batchCounts: any;
```

Comment:
would be great to have a `type BatchCounts` with a comment on what's that for

Reply:
Ah yes, so for _innerBulk it would be a node driver type but the shell-api class can't import that directly. So instead you are saying we should define interfaces in the service-provider-core that indicate what the innerBulk/batches types should be? I can definitely do that